### PR TITLE
graph/multi: clean up missed changes and bugs

### PR DESCRIPTION
**Pull request by kortschak**
This fixes some bugs in the multigraph handling of edges from lines and finishes changes that I had intended to go into the iterator PR (the change in container for `graph.*Lines` in the `multi.*Edge` types).

It has also raised an issue of how I have been dealing with empty collections. With the old approach we could always return a `nil` and people could still iterate over it without needing to check for `nil`. This is no longer true. We currently don't specify the `nil`-ness of returns for collections, so this is not terrible, but it might be an idea to tighten that down. I'm raising this here since you can see the impacts with these changes, but I will file an issue (#614).

Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
Updates #31.

**Comment by kortschak** [comment]
*ping*

File: graph/multi/multi.go [hunk 1]
```diff
@@ -54,41 +48,42 @@ func (l Line) ID() int64 { return l.UID }
...
 // Weight returns the weight of the edge.
 func (e WeightedEdge) Weight() float64 {
+	if e.WeightedLines == nil {
+		return 0
+	}
 	if e.WeightFunc == nil {
 		var w float64
-		for _, l := range e.Lines {
-			w += l.Weight()
+		for e.Next() {
```
Comment (vladimir-ch) [inline graph/multi/multi.go]: Is it ok to assume that `e` is at the first line? Above it is documented for the input to `WeightFunc` but what about `Weight`?

File: graph/multi/weighted_directed.go [hunk 2]
```diff
@@ -30,7 +30,11 @@ var (
 
 // WeightedDirectedGraph implements a generalized directed graph.
 type WeightedDirectedGraph struct {
-	EdgeWeightFunc func([]graph.WeightedLine) float64
+	// EdgeWEightFunc is used to provide
```
Comment (vladimir-ch) [inline graph/multi/weighted_directed.go]: `EdgeWeightFunc` (lower-case e)

File: graph/multi/directed.go [hunk 3]
```diff
@@ -181,15 +181,21 @@ func (g *DirectedGraph) Nodes() graph.Nodes {
...
 		for _, e := range g.from[u.ID()] {
-			var lines Edge
+			var lines []graph.Line
 			for _, l := range e {
 				lines = append(lines, l)
 			}
 			if len(lines) != 0 {
-				edges = append(edges, lines)
+				edges = append(edges, Edge{
+					F: g.Node(u.ID()), T: g.Node(lines[0].To().ID()),
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/multi/directed.go]: Changed

File: graph/multi/multi.go (hunk 1 above)
Comment (kortschak) [reply graph/multi/multi.go]: Added documentation making this clear.

File: graph/multi/weighted_directed.go (hunk 2 above)
Comment (kortschak) [reply graph/multi/weighted_directed.go]: Thanks. Fixed.

**Comment by kortschak** [comment]
PTAL

**Pull request merged**